crypto/src/proofs/transparent.rs

```diff
@@ -418,57 +436,109 @@ impl SwapClaimProof {
+        // check the swap NFT Asset ID is properly constructed
+        let asset_id = self.swap_nft_asset_id;
+        let expected_asset_id = swap::generate_swap_asset_id(
+            self.delta_1,
+            self.delta_2,
+            fee,
+            self.b_d,
+            self.pk_d,
+            self.trading_pair,
+        )
+        .map_err(|_| anyhow!("error generating expected swap asset ID"))?;
+        if expected_asset_id != asset_id {
+            return Err(anyhow!("improper swap NFT asset id"));
+        }
 
         // Merkle path integrity.
-        self.note_commitment_block_proof
+        self.note_commitment_proof
             .verify(anchor)
             .map_err(|_| anyhow!("merkle root mismatch"))?;
 
-        // Value commitment integrity.
-        if self.swap_nft_value.commit(self.v_blinding) != value_commitment {
-            return Err(anyhow!("value commitment mismatch"));
+        // Validate the note commitment was for the proper block height.
+        let position = self.note_commitment_proof.position();
+        let block = position.block();
+        let epoch = position.epoch();
+        let note_commitment_block_height: u64 =
+            epoch_duration * u64::from(epoch) + u64::from(block);
+        if note_commitment_block_height != clearing_price_height {
+            return Err(anyhow::anyhow!(
+                "note commitment was not for clearing price height"
+            ));
+        }
+
+        // Confirms that `value_commitment` is a commitment to `fee` penumbra tokens
+        let fee_value = Value {
+            amount: fee,
+            asset_id: *STAKING_TOKEN_ASSET_ID,
+        };
+        // Fees are public and use a 0 blinding factor
+        let fee_v_blinding = Fr::zero();
+        let expected_fee_value_commitment = fee_value.commit(fee_v_blinding);
+        if expected_fee_value_commitment != value_commitment {
+            return Err(anyhow!("fee value commitment mismatch"));
         }
 
         // Nullifier integrity.
         if nullifier
-            != self.nk.derive_nullifier(
-                self.note_commitment_position,
-                &self.note_commitment_block_proof.commitment(),
-            )
+            != self
+                .nk
+                .derive_nullifier(position, &self.note_commitment_proof.commitment())
         {
             return Err(anyhow!("bad nullifier"));
         }
 
-        // TODO: is there an address check that needs to occur?
+        // TODO:
+        // The address should be the same for the Swap NFT and SwapClaim outputs
+        // Need output notes here, and to validate the amounts and addresses.
+        // let lambda_1 = success.into() * (clearing_price_1 * self.delta_2)
+        //     + (1 - success.into()) * self.delta_1;
+        // let lambda_2 = success.into() * (clearing_price_2 * self.delta_1)
+        //     + (1 - success.into()) * self.delta_2;
```

Comment:
We should also tweak the description in the protocol spec to match the data defined in #1126 / #1214, so that the formulas in the spec are written in terms of a pro rata share of the batch, rather than in terms of a multiplication-by-price-factor (which is harder to do precisely).